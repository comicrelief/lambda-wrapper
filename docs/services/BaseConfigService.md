# BaseConfigService

Instead of reimplementing the service status get and set logic across several services, Lambda Wrapper provides a Status service that handles these two operations for us.

## Usage

This class is to be extended by the implementing services so that `defaultConfig` and possibly `s3Config` can be overriden / extended. As such, it is not included as a dependency by default and must be explicitly added.

Example implementation with validation:

```ts
// src/services/ConfigService.ts
import { BaseConfigService } from '@comicrelief/lambda-wrapper';

import { ConfigModel, ConfigProps } from '@/src/models/Config';

export default class ConfigService extends BaseConfigService {
  async put(config): Promise<ConfigProps> {
    const validated = await ConfigModel.validate(config);
    return super.put(validated);
  }

  async get(): Promise<ConfigProps> {
    const config = await super.get();
    return ConfigModel.validate(config);
  }
}
```

Config is typed as `unknown` in the base class since you shouldn't trust what's in the bucket. Override the `get` and `put` methods to pass the results through some validation to ensure the config is valid and can safely be typed.

Then add to your Lambda Wrapper dependencies:

```ts
// src/config/lambda-wrapper.ts
import lambdaWrapper from '@comicrelief/lambda-wrapper';

import ConfigService from '@/src/services/ConfigService';

export default lambdaWrapper.configure({
  dependencies: {
    ConfigService,
    // ...
  },
});
```
