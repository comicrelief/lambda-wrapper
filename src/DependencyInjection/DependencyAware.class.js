/**
 * DependencyAwareClass Class
 */
export default class DependencyAwareClass {
  /**
   * DependencyAwareClass constructor
   * @param di DependencyInjection
   */
  constructor(di) {
    this.di = di;
  }

  /**
   * Get Dependency Injection Container
   * @return DependencyInjection
   */
  getContainer() {
    return this.di;
  }
}
