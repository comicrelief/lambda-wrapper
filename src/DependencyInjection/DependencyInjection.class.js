import { DEFINITIONS, DEPENDENCIES } from '../Config/Dependencies';

/**
 * DependencyInjection class
 */
export default class DependencyInjection {
  /**
   * DependencyInjection constructor
   * @param configuration
   * @param event
   * @param context
   */
  constructor(configuration, event, context) {
    this.event = event;
    this.context = context;

    this.dependencies = {};
    this.configuration = configuration;

    for (let x = 0; x <= 1; x += 1) {
      // Iterate over lapper dependencies and add to container
      Object.keys(DEFINITIONS).forEach((dependencyKey) => {
        this.dependencies[dependencyKey] = new DEPENDENCIES[dependencyKey](this);
      });

      // Iterate over child dependencies and add to container
      if (typeof configuration.DEPENDENCIES !== 'undefined') {
        Object.keys(configuration.DEPENDENCIES).forEach((dependencyKey) => {
          this.dependencies[dependencyKey] = new configuration.DEPENDENCIES[dependencyKey](this);
        });
      }
    }
  }

  /**
   * Get Dependency
   * @param definition
   * @return {*}
   */
  get(definition) {
    if (typeof this.dependencies[definition] === 'undefined') {
      throw Error(`${definition} does not exist in di container`);
    }

    return this.dependencies[definition];
  }

  /**
   * Get Event
   * @return {*}
   */
  getEvent() {
    return this.event;
  }

  /**
   * Get Context
   * @return {*}
   */
  getContext() {
    return this.context;
  }

  /**
   * Get Configuration
   * @param definition string
   * @return {*}
   */
  getConfiguration(definition = null) {
    if (definition !== null && typeof this.configuration[definition] === 'undefined') {
      return null;
    } else if (typeof this.configuration[definition] !== 'undefined') {
      return this.configuration[definition];
    }

    return this.configuration;
  }
}
