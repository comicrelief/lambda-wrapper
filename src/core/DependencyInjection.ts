import { Context } from 'aws-lambda';

import DependencyAwareClass from './DependencyAwareClass';
import { LambdaWrapperConfig } from './config';

// eslint-disable-next-line no-use-before-define
type Class<TInstance, TConfig extends LambdaWrapperConfig> = new (di: DependencyInjection<TConfig>) => TInstance;

/**
 * Dependency injection container.
 *
 * Dependencies (singleton instances of dependency-aware classes) are provided
 * to the main Lambda handler and other dependencies via this class.
 */
export default class DependencyInjection<TConfig extends LambdaWrapperConfig = any> {
  /**
   * Instantiated dependencies.
   */
  readonly dependencies: Record<string, DependencyAwareClass>;

  /**
   * True until all dependencies have been constructed.
   */
  private isConstructing = true;

  constructor(
    readonly config: TConfig,
    readonly event: any,
    readonly context: Context,
  ) {
    // get unique dependency classes -- a class may be included several times,
    // but should be instantiated only once
    const classes = Array.from(new Set(Object.values(config.dependencies)));

    // guard against duplicate keys
    const countByName = classes
      .map((Constructor) => Constructor.name)
      .reduce(
        (counts, name) => ({ ...counts, [name]: (counts[name] || 0) + 1 }),
        {} as Record<string, number>,
      );
    if (Object.values(countByName).some((count) => count > 1)) {
      const duplicateNames = Object.entries(countByName)
        .filter(([, count]) => count > 1)
        .map(([name]) => name);

      // if all class names are single-letter, they're probably minified -- in
      // this case, give a hint about how to fix it
      const action = duplicateNames.every((it) => it.length === 1)
        ? "If you don't recognise the single-letter names listed above, your "
          + "bundler may be minifying your code. You'll need to disable this "
          + 'for Lambda Wrapper to work correctly. Please refer to the Notes '
          + 'section of the Lambda Wrapper readme:\n\n'
          + '  https://github.com/comicrelief/lambda-wrapper#notes'
        : 'Please ensure that all dependency classes have a unique name.';

      throw new Error(
        `Dependency names are not unique: ${duplicateNames.join(', ')}\n\n${action}`,
      );
    }

    // instantiate all dependencies
    this.dependencies = Object.fromEntries(
      classes.map((Constructor) => [Constructor.name, new Constructor(this)]),
    );

    this.isConstructing = false;
  }

  /**
   * Get the singleton instance of the given dependency.
   *
   * @param dependency
   */
  get<T extends DependencyAwareClass>(dependency: Class<T, TConfig>): T {
    if (this.isConstructing) {
      throw new Error(
        'Dependencies are not available in dependency class constructors.\n\n'
        + 'To fix this, call `di.get` in the function where the dependency is '
        + 'used instead of inside your constructor.',
      );
    }

    const name = dependency.name;

    if (!this.dependencies[name]) {
      throw new Error(
        `${name} does not exist in dependency container\n\n`
        + `Make sure you've included ${name} in the 'dependencies' key of your `
        + 'Lambda Wrapper config.',
      );
    }

    return this.dependencies[name] as T;
  }

  /**
   * Get the event passed to AWS Lambda.
   *
   * @deprecated Use `di.event` instead.
   */
  getEvent() {
    return this.event;
  }

  /**
   * Get the AWS Lambda context object.
   *
   * @deprecated Use `di.context` instead.
   */
  getContext() {
    return this.context;
  }

  /**
   * Get Lambda Wrapper configuration.
   *
   * @deprecated Use `di.config` instead.
   */
  getConfiguration() {
    return this.config;
  }

  /**
   * True if the function is being executed in `serverless-offline`.
   *
   * We use the following checks for this:
   *
   * - if there is no function ARN, or the ARN includes 'offline'
   * - if `process.env.USE_SERVERLESS_OFFLINE` is set
   *
   * TODO: This is nothing to do with dependency injection and should be moved
   *   somewhere else! Any ideas?
   */
  get isOffline(): boolean {
    return !this.context.invokedFunctionArn
      || this.context.invokedFunctionArn.includes('offline')
      || !!process.env.USE_SERVERLESS_OFFLINE;
  }

  /**
   * Get the `service-stage-` prefix added by Serverless to deployed Lambda
   * function names. This is handy when you want to invoke other functions,
   * without having to hardcode the service name and stage.
   *
   * The returned prefix includes a trailing dash. To get the deployed name of
   * another Lambda function, concatenate its serverless function name (its key
   * in `serverless.yml`) onto the prefix:
   *
   * ```js
   * const serverlessFunctionName = 'MyFunction';
   * const deployedName = `${di.getLambdaPrefix()}${serverlessFunctionName}`;
   * ```
   *
   * This function relies on looking at the currently running Lambda function's
   * resource name. It will not work correctly if the Lambda function has been
   * given a custom resource name.
   */
  getLambdaPrefix(): string {
    const stage = process.env.STAGE;
    if (!stage) {
      /* eslint-disable no-template-curly-in-string */
      throw new Error(
        'STAGE is not set\n\n'
        + 'Please add to your Lambda environment:\n\n'
        + '    STAGE: ${sls:stage}\n',
      );
    }
    if (!this.context.functionName) {
      throw new Error('Lambda function name is unavailable in context');
    }
    const [service] = this.context.functionName.split(`-${stage}-`);
    return `${service}-${stage}-`;
  }
}
