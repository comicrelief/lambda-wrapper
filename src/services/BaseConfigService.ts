import {
  DeleteObjectCommand,
  GetObjectCommand,
  PutObjectCommand,
  S3Client,
} from '@aws-sdk/client-s3';

import DependencyAwareClass from '../core/DependencyAwareClass';
import LambdaTermination from '../utils/LambdaTermination';

/**
 * `error.Code` for S3 404 errors.
 */
export const S3_NO_SUCH_KEY_ERROR_CODE = 'NoSuchKey';

/**
 * Represents the service states.
 */
export const ServiceStates = {
  OK: 'OK',
  TEMPORARILY_PAUSED: 'TEMPORARILY_PAUSED',
  INDEFINITELY_PAUSED: 'INDEFINITELY_PAUSED',
};

/**
 * Maps service states to HTTP codes.
 */
export const ServiceStatesHttpCodes = {
  [ServiceStates.OK]: 200,
  [ServiceStates.TEMPORARILY_PAUSED]: 409,
  [ServiceStates.INDEFINITELY_PAUSED]: 409,
};

/**
 * This class is to be extended by the implementing services so that
 * `defaultConfig` and possibly `s3Config` can be overriden / extended.
 *
 * Config is typed as `unknown` since you shouldn't trust what's in the bucket.
 * Override the `get` and `put` methods to pass the results through some
 * validation to ensure the config is valid and can safely be typed.
 */
export default class BaseConfigService extends DependencyAwareClass {
  /**
   * Returns the basic config.
   *
   * This getter is used to set the default config should the service not find
   * any on the configured S3 Bucket.
   *
   * See `getOrCreate` and `patch` methods.
   */
  static get defaultConfig() {
    return {
      state: ServiceStates.OK,
    };
  }

  /**
   * Returns the S3 configuration used to retrieve or update the service
   * configuration.
   */
  static get s3config(): { Bucket: string; Key: string; } {
    return {
      Bucket: process.env.SERVICE_CONFIG_S3_BUCKET || '',
      Key: process.env.SERVICE_CONFIG_S3_KEY || '',
    };
  }

  /**
   * Returns an S3 client.
   */
  static get client() {
    return new S3Client({
      region: process.env.REGION,
    });
  }

  /**
   * Returns an S3 client
   */
  get client() {
    return (this.constructor as typeof BaseConfigService).client;
  }

  /**
   * Deletes the configuration stored on S3. Helpful in feature tests.
   */
  async delete() {
    return this.client.send(new DeleteObjectCommand(
      (this.constructor as typeof BaseConfigService).s3config,
    ));
  }

  /**
   * Puts the given configuration on S3.
   *
   * @param config
   */
  async put<T>(config: T): Promise<T> {
    await this.client.send(new PutObjectCommand({
      ...(this.constructor as typeof BaseConfigService).s3config,
      Body: JSON.stringify(config),
    }));

    return config;
  }

  /**
   * Gets the service configuration.
   */
  async get(): Promise<unknown> {
    const response = await this.client.send(new GetObjectCommand(
      (this.constructor as typeof BaseConfigService).s3config,
    ));
    const body = await response.Body?.transformToString();

    if (!body) {
      // Empty strings are not valid configurations
      throw new Error('Configuration file is empty');
    }

    try {
      return JSON.parse(body);
    } catch {
      throw new Error('Invalid configuration file');
    }
  }

  /**
   * Gets or creates the service configuration.
   *
   * If the configuration is not found on S3 the default configuration is
   * uploaded and returned instead.
   */
  async getOrCreate(): Promise<unknown> {
    try {
      return await this.get();
    } catch (error: any) {
      if (error.Code !== S3_NO_SUCH_KEY_ERROR_CODE) {
        // Throw directly any other error
        throw error;
      }

      return this.put((this.constructor as typeof BaseConfigService).defaultConfig);
    }
  }

  /**
   * Patches the existing configuration
   * or the default configuration
   * with the provided partial configuration
   *
   * @param partialConfig
   */
  async patch(partialConfig: any): Promise<unknown> {
    let base: any = (this.constructor as typeof BaseConfigService).defaultConfig;

    try {
      base = await this.get();
    } catch (error: any) {
      if (error.Code !== S3_NO_SUCH_KEY_ERROR_CODE) {
        // Throw directly any other error
        throw error;
      }

      // Config doesn't exist
      // keep using `this.constructor.defaultConfig`
    }

    return this.put({
      ...base,
      ...partialConfig,
    });
  }

  /**
   * Performs a health check given the current config.
   *
   * If `currentConfig` is not supplied it uses `getOrCreate` to fetch it.
   *
   * @param currentConfig
   */
  async healthCheck(currentConfig?: any): Promise<number> {
    const config = currentConfig || await this.getOrCreate();

    return ServiceStatesHttpCodes[config.state] || 500;
  }

  /**
   * Ensures that the application is healthy or throws a `LambdaTermination`.
   *
   * @param currentConfig
   */
  async ensureHealthy(currentConfig: any = null) {
    const statusCode = await this.healthCheck(currentConfig);

    if (statusCode < 400) {
      return statusCode;
    }

    const message = 'Application is not healthy.';

    throw new LambdaTermination(message, statusCode, message, message);
  }
}
