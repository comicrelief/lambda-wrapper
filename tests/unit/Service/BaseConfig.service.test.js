import { S3 } from 'aws-sdk';

import DependencyInjection from '../../../src/DependencyInjection/DependencyInjection.class';
import BaseConfigService, { S3_NO_SUCH_KEY_ERROR_CODE, ServiceStates, ServiceStatesHttpCodes } from '../../../src/Service/BaseConfig.service';

const createAsyncMock = (returnValue) => {
  const mockedValue = returnValue instanceof Error
    ? Promise.reject(returnValue)
    : Promise.resolve(returnValue);

  return jest.fn().mockReturnValue({ promise: () => mockedValue });
};

/**
 * Generates a BaseConfigService
 * @param {*} param0
 * @returns {BaseConfigService}
 */
const getService = ({ getObject = null, putObject = null, deleteObject = null } = {}) => {
  const di = new DependencyInjection({}, {}, {});
  const service = new BaseConfigService(di);
  const client = {
    getObject: createAsyncMock(getObject),
    putObject: createAsyncMock(putObject),
    deleteObject: createAsyncMock(deleteObject),
  };

  jest.spyOn(service, 'client', 'get').mockReturnValue(client);

  return service;
};

const BaseConfigUnitTests = (serviceGenerator: (...args) => BaseConfigService) => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('defaultConfig', () => {
    it('is a valid object', () => {
      const service = serviceGenerator();
      const isValidObject = typeof service.constructor.defaultConfig === 'object' && service.constructor.defaultConfig !== null;

      expect(isValidObject).toEqual(true);
    });

    it('has state defined', () => {
      const service = serviceGenerator();
      const defaultConfig = service.constructor.defaultConfig;

      expect('state' in defaultConfig).toEqual(true);
    });
  });

  describe('s3config', () => {
    it('is a valid object', () => {
      const service = serviceGenerator();
      const isValidObject = typeof service.constructor.s3config === 'object' && service.constructor.s3config !== null;

      expect(isValidObject).toEqual(true);
    });

    it('has Bucket and Key defined', () => {
      const service = serviceGenerator();
      const s3config = service.constructor.s3config;

      expect('Bucket' in s3config).toEqual(true);
      expect('Key' in s3config).toEqual(true);
    });
  });

  describe('delete', () => {
    it('calls client.deleteObject', async () => {
      const service = serviceGenerator();
      await service.delete();

      expect(service.client.deleteObject).toHaveBeenCalledTimes(1);
    });
  });

  describe('put', () => {
    it('calls client.putObject', async () => {
      const expected = Symbol('put');
      const service = serviceGenerator();
      await service.put(expected);

      expect(service.client.putObject).toHaveBeenCalledTimes(1);
    });

    it('returns the provided config unchanged', async () => {
      const expected = Symbol('put');
      const service = serviceGenerator();
      const config = await service.put(expected);

      expect(config).toEqual(expected);
    });
  });

  describe('get', () => {
    it('gets an existing config', async () => {
      const expected = { a: 1 };
      const service = serviceGenerator({ getObject: { Body: JSON.stringify(expected) } });
      const config = await service.get();

      expect(config).toEqual(expected);
    });

    it('refuses empty configurations', async () => {
      const service = serviceGenerator({ getObject: { Body: '' } });

      await expect(service.get()).rejects.toThrowErrorMatchingSnapshot();
    });

    it('refuses invalid configurations', async () => {
      const service = serviceGenerator({ getObject: { Body: '{ "a": 1' } });

      await expect(service.get()).rejects.toThrowErrorMatchingSnapshot();
    });

    it('propagates the 404', async () => {
      const error = new Error('404');
      error.code = S3_NO_SUCH_KEY_ERROR_CODE;

      const service = serviceGenerator({ getObject: error });

      await expect(service.get()).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('getOrCreate', () => {
    it('uploads the defaultConfig with a 404 error', async () => {
      const error = new Error('404');
      error.code = S3_NO_SUCH_KEY_ERROR_CODE;

      const service = serviceGenerator({ getObject: error });
      const config = await service.getOrCreate();

      expect(config).toEqual(service.constructor.defaultConfig);
    });

    it('throws any non-404 error', async () => {
      const error = new Error('Bad error');
      error.code = 'another';

      const service = serviceGenerator({ getObject: error });

      await expect(service.getOrCreate()).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('patch', () => {
    it('uses the existing config if an existing config is found', async () => {
      const existing = { a: 1 };
      const service = serviceGenerator({ getObject: { Body: JSON.stringify(existing) } });

      const additional = { b: 2 };
      const expected = { ...existing, ...additional };
      const config = await service.patch(additional);

      expect(config).toEqual(expected);
    });

    it('uses the base config if no existing config is found', async () => {
      const error = new Error('404');
      error.code = S3_NO_SUCH_KEY_ERROR_CODE;
      const service = serviceGenerator({ getObject: error });

      const existing = service.constructor.defaultConfig;
      const additional = { b: 2 };
      const expected = { ...existing, ...additional };
      const config = await service.patch(additional);

      expect(config).toEqual(expected);
    });

    it('throws any non-404 error', async () => {
      const error = new Error('Bad error');
      error.code = 'another';

      const service = serviceGenerator({ getObject: error });

      await expect(service.patch({ b: 1 })).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('healthCheck', () => {
    Object.values(ServiceStates).forEach((state) => {
      describe(state, () => {
        it('Returns the expected HTTP code with the given config', async () => {
          const config = { state };
          const service = serviceGenerator();
          const statusCode = await service.healthCheck(config);
          const expected = ServiceStatesHttpCodes[state];

          expect(statusCode).toEqual(expected);
        });

        it('Returns the expected HTTP code with the existing config', async () => {
          const config = { state };
          const service = serviceGenerator({ getObject: { Body: JSON.stringify(config) } });
          const statusCode = await service.healthCheck();
          const expected = ServiceStatesHttpCodes[state];

          expect(statusCode).toEqual(expected);
        });
      });
    });

    describe('Unknown state', () => {
      it('Returns 500 with the given config', async () => {
        const config = { state: 'Unknown' };
        const service = serviceGenerator();
        const statusCode = await service.healthCheck(config);
        const expected = 500;

        expect(statusCode).toEqual(expected);
      });

      it('Returns 500 with the existing config', async () => {
        const config = { state: 'Unknown' };
        const service = serviceGenerator({ getObject: { Body: JSON.stringify(config) } });
        const statusCode = await service.healthCheck();
        const expected = 500;

        expect(statusCode).toEqual(expected);
      });
    });
  });

  describe('ensureHealthy', () => {
    [200, 201, 202, 204, 300, 301, 399].forEach((statusCode) => {
      describe(statusCode, () => {
        it('is healthy', async () => {
          const service = serviceGenerator();
          jest.spyOn(service, 'healthCheck').mockImplementation(() => Promise.resolve(statusCode));

          await expect(service.ensureHealthy()).resolves.toEqual(statusCode);
        });
      });
    });

    [400, 401, 403, 404, 409, 499, 500, 501, 502, 503, 504, 'Dante Alighieri'].forEach((statusCode) => {
      describe(statusCode, () => {
        it('throws a LambdaTermination', async () => {
          const service = serviceGenerator();
          jest.spyOn(service, 'healthCheck').mockImplementation(() => Promise.resolve(statusCode));

          await expect(service.ensureHealthy()).rejects.toThrowErrorMatchingSnapshot();
        });
      });
    });
  });
};

describe('Service/BaseConfigService', () => {
  BaseConfigUnitTests(getService);

  describe('client', () => {
    it('Returns an s3 instance (static)', () => {
      expect(BaseConfigService.client instanceof S3).toEqual(true);
    });

    it('Returns an s3 instance', () => {
      const di = new DependencyInjection({}, {}, {});
      const service = new BaseConfigService(di);

      expect(service.client instanceof S3).toEqual(true);
    });
  });
});
