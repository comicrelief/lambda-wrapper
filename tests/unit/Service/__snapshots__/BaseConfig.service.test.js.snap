// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service/BaseConfigService get propagates the 404 1`] = `"404"`;

exports[`Service/BaseConfigService get refuses empty configurations 1`] = `"Configuration file is empty"`;

exports[`Service/BaseConfigService get refuses invalid configurations 1`] = `"Invalid configuration file"`;

exports[`Service/BaseConfigService getOrCreate throws any non-404 error 1`] = `"Bad error"`;

exports[`Service/BaseConfigService patch throws any non-404 error 1`] = `"Bad error"`;
